addons/l10n_id_efaktur_coretax/models/account_move.py

```diff
@@ -0,0 +1,317 @@
+            if not record.company_id.vat:
+                raise ValidationError(_("Please configre the VAT of your company"))
+
+        # Should prevent users from generating e-Faktur document on invoices across multi-company.
+        # Allowing it will cause issues on the invoice/eFaktur document record rule
+        if len(self.company_id) > 1:
+            raise UserError(_("You are not allowed to generate e-Faktur document from invoices coming from different companies"))
+
+        # All invoices in self have no documents; we can create a new one for them.
+        # Or all invoices in self have a document, but it's the same one. Special use case but we allow downloading it.
+        if not self.l10n_id_efaktur_document:
+            self.l10n_id_efaktur_document = self.env['l10n_id_efaktur_coretax.document'].create({
+                'invoice_ids': self.ids,
+                'company_id': self.company_id.id,
+            })
+            self.l10n_id_efaktur_document.action_regenerate()
+
+        # If there is more than one document, or all invoices for a document were not selected, the resulting file could cause mistakes;
+        # They could get a file with additional invoices for example. In this case, we redirect them to the document view to make it clearer.
+        elif len(self.l10n_id_efaktur_document) > 1 or set(self.l10n_id_efaktur_document.invoice_ids.ids) != set(self.ids):
+            action_error = {
+                'name': _('Document Mismatch'),
+                'view_mode': 'list',
+                'res_model': 'l10n_id_efaktur_coretax.document',
+                'type': 'ir.actions.act_window',
+                'views': [[False, 'list'], [False, 'form']],
+                'domain': [('id', 'in', self.l10n_id_efaktur_document.ids)],
+            }
+            msg = _("The selected invoices are partially part of one or more e-faktur documents.\n"
+                    "Please download them from the e-faktur documents directly.")
+            raise RedirectWarning(msg, action_error, _("Display Related Documents"))
+
+        return self.download_xml()
+
+    def download_xml(self):
+        return self.l10n_id_efaktur_document.action_download()
+
+    def prepare_efaktur_vals(self):
+        """ Get information required from invoice and lines to generate E-Faktur that will be used
+        to load in the XML template later on"""
+        invoice_vals = []
+        idr = self.env.ref('base.IDR')
+
+        for move in self.filtered(lambda m: m.state == 'posted'):
+
+            commercial_partner = move.partner_id.commercial_partner_id
+            trx_code = move.l10n_id_kode_transaksi
+
+            vals = {
+                "TIN": move.company_id.vat,
+                "TaxInvoiceDate": move.invoice_date.strftime("%Y-%m-%d"),
+                "TaxInvoiceOpt": "Normal",
+                "TrxCode": trx_code,
+                "AddInfo": move['l10n_id_add_info_' + trx_code] if trx_code in ('07', '08') else "",
```

Comment:
I'd prefer if you did an if/else instead of dynamically accessing the move data like this.
SAme thing for FacilityStamp.